### PR TITLE
Set all serialised date parts at once

### PR DESCRIPTION
**Pull request by Vinnl**
Previously, we set each date part individually (i.e. first the
year, then the month, etc.). However, if any of these parts result
in an invalid/unrepresentable date (e.g. because the year is
earlier than the earliest year that can be represented by
JavaScript's Date object), the end result will also be invalid,
even though the parts we set later would make them valid again
(e.g. because the timezone pushes the date back into the range that
is representable by Date).

To fix this, this commit sets all Date parts at the same time.

This bug was discovered by our property-based tests, which tested more edge cases with fast-check's new version. Hence, this PR should make #609 mergeable.

- [x] I've added a unit test to test for potential regressions of this bug.
- [x] The changelog has been updated, if applicable.
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

### CODE REVIEW

File: src/datatypes.test.ts [hunk 1]
```diff
@@ -224,6 +224,44 @@ describe("deserializeDatetime", () => {
     expect(deserializeDatetime("1990-11-12T10:00:00+14:00")).toEqual(
       expectedDateWithMaxPositiveTimezone
     );
+
+    const expectedEarliestRepresentableDate = new Date(-8640000000000000);
```
Comment (pmcb55) [inline src/datatypes.test.ts:228]: I'd just make this (and the positive version) file-level constants with descriptive names - i.e. the same values, plus and minus, are used in a couple of different contexts (i.e. in Earliest, Latest, Earlier Timezone, in Later Timezone), e.g.
```
const EARLIEST_POSSIBLE_VALID_DATE = new Date(-8640000000000000);
const LATEST_POSSIBLE_VALID_DATE = new Date(8640000000000000);
```

**Review by pmcb55 — APPROVED** [review]
LGTM

File: src/datatypes.test.ts (hunk 1 above)
Comment (Vinnl) [reply src/datatypes.test.ts:228]: Huh, good point, those shouldn't have been different variables - the point of those tests is to represent the _same_ value with different serialisations. Fixed now.
